bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/test/java/com/google/cloud/bigtable/mapreduce/validation/EndToEndIT.java

```diff
@@ -0,0 +1,429 @@
+    htd.addFamily(columnDesc);
+    admin.createTable(htd, generateSplits(numRows, numRegions));
+
+    LOG.info("hbase table created: " + htd.toString());
+
+    return HB_TEST_UTIL.getConnection().getTable(table);
+  }
+
+  private static byte[][] generateSplits(int numRows, int numRegions) {
+    byte[][] splitRows = new byte[numRegions - 1][];
+    for (int i = 1; i < numRegions; i++) {
+      splitRows[i - 1] = Bytes.toBytes(numRows * i / numRegions);
+    }
+    return splitRows;
+  }
+
+  public static String getZkConnArgFromConfig(Configuration conf) {
+    return new StringBuilder()
+        .append(conf.get(HConstants.ZOOKEEPER_QUORUM))
+        .append(":")
+        .append(conf.get(HConstants.ZOOKEEPER_CLIENT_PORT))
+        .append(":")
+        .append(conf.get(HConstants.ZOOKEEPER_ZNODE_PARENT))
+        .toString();
+  }
+
+  /** validate data migrated from Bigtable to HBase */
+  @Test
+  public void testSyncTableBigtabletoHBaseValidation() throws Exception {
+    long testRunTs = System.currentTimeMillis();
+
+    // source and target
+    TableName BT_TEST_SOURCE_TABLE = TableName.valueOf("cbt-testsourcetable-" + testRunTs);
+    TableName HB_TEST_TARGET_TABLE = TableName.valueOf("hbase-testtargettable-" + testRunTs);
+    Path testHashOutDir = HB_TEST_UTIL.getDataTestDirOnTestFS("hash-out-cbt-" + testRunTs);
+
+    // generate data
+    writeTestData(BT_TEST_SOURCE_TABLE, false, HB_TEST_TARGET_TABLE, true);
+
+    // hash table
+    hashSourceBigtable(BT_TEST_SOURCE_TABLE, testHashOutDir);
+
+    // sync table validation
+    Counters syncCounters =
+        bigtableSyncTableJobSourceBigtableTargetHBase(
+            BT_TEST_SOURCE_TABLE, HB_TEST_TARGET_TABLE, testHashOutDir);
+
+    assertVerifySyncMatches(syncCounters);
+  }
+
+  /** validate data migrated from HBase to Bigtable */
+  @Test
+  public void testSyncTableHBaseToBigtableValidation() throws Exception {
+    long testRunTs = System.currentTimeMillis();
+
+    // source and target
+    TableName HB_TEST_SOURCE_TABLE = TableName.valueOf("hbase-testsourcetable-" + testRunTs);
+    TableName BT_TEST_TARGET_TABLE = TableName.valueOf("cbt-testtargettable-" + testRunTs);
+    Path testHashOutDir = HB_TEST_UTIL.getDataTestDirOnTestFS("hash-out-hbase-" + +testRunTs);
+
+    // generate data
+    writeTestData(HB_TEST_SOURCE_TABLE, true, BT_TEST_TARGET_TABLE, false);
+
+    // hash table
+    hashSourceHBaseTable(HB_TEST_SOURCE_TABLE, testHashOutDir);
+
+    // sync table validation
+    Counters syncCounters =
+        bigtableSyncTableJobSourceHBaseTargetBigtable(
+            HB_TEST_SOURCE_TABLE, BT_TEST_TARGET_TABLE, testHashOutDir);
+
+    assertVerifySyncMatches(syncCounters);
+  }
+
+  /** assert match/mismatch results from counters and output records */
+  private void assertVerifySyncMatches(Counters syncCounters) {
+    // verify counters
+    assertEquals(60, syncCounters.findCounter(Counter.ROWSWITHDIFFS).getValue());
```

Comment:
There counters don't match? 

RowsWithDiff = Sum (all other diffs)??

Reply:
NIT: The whole setup is hidden away somewhere, the readability of this test can be better. Its hard to comprehend the magic numbers that you are validating.